In [11]:
import pandas as pd
import numpy as np
import seaborn as sns

# Load Penguins
penguins = sns.load_dataset("penguins") 

# Quick Look
penguins.head()

#How many rows and columns?
penguins.shape

#Hwhat are the columns?
penguins.columns

# Data counts and null values
penguins.info()
penguins.isna().sum()

# Summary Statistics
penguins.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
count,342.000000,342.000000,342.000000,342.000000
mean,43.921930,17.151170,200.915205,4201.754386
std,5.459584,1.974793,14.061714,801.954536
min,32.100000,13.100000,172.000000,2700.000000
25%,39.225000,15.600000,190.000000,3550.000000
50%,44.450000,17.300000,197.000000,4050.000000
75%,48.500000,18.700000,213.000000,4750.000000
max,59.600000,21.500000,231.000000,6300.000000


### IQR = Interquartile range
- 25% Lower quartile (Q1), 25% of values fall below this number
- 50% Median middle value
- 75% Upper quartile (Q3), 75% of values rise above this number

### IQR = Q3 - Q1 -> the spread of the 'middle 50%' of data
- Outlier are usually flagged using the 1.5 * IQR rule:
- Lower bound = Q1 - (1.5*IQR)
- Upper bound = Q3 + (1.5*IQR)
- Anything outside these are considered outliers

### Example: body_mass_g 
- Q1 = 3550, Q3 = 4750
- IQR = (Q3 - Q1) = 1200
- What is 1.5 * 1200? 1800
- Lower bound = 3550 - 1800 = 1750
- Higher bound = 4750 + 1800 = 6550

### Example: bill_length_mm
- Q1 = 48.5, Q3 = 39.2
- IQR = (Q3 - Q1) = 9.3
- What is 1.5 * 9.3? 13.95
- Lower bound = 39.2 - 13.95 = 25.25
- Higher bound = 48.5 + 13.95 = 62.45




In [ ]:
# Categorical overview: how many of each category
for col in ["species", "island", "sex"]:
    print(f"\n== {col} ==")
    print(penguins[col].value_counts(dropna=False))  # include NaN to see

In [12]:
#Make a working copy to clean
penguins_clean = penguins.copy()

#Standardise column names(tidy, snake case)
penguins_clean.columns = (
penguins_clean.columns.str.strip() #Remove whitespace before and after
.str.lower() #lowercase
.str.replace(" ", "_") #replace spaces with underscores
)


#Fix/declair categorical data types where appropriate
cat_cols = ["species", "island", "sex"]
for col in cat_cols:
    penguins_clean[col] = penguins_clean[col].astype("category") #Saves memory and clarifies intent

penguins_clean.info() #Check changes




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     342 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  342 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                333 non-null    category
dtypes: category(3), float64(4)
memory usage: 12.3 KB


In [ ]:
# Deal with numerical data